hera_sim/beams.py

```diff
@@ -0,0 +1,289 @@
+        if self.xstretch != 1. or self.ystretch != 1.:
+            # Convert sheared Cartesian coords to circular polar coords
+            # mX stretches in x direction, mY in y direction, a is angle
+            # Notation: phi = az, theta = za. Subscript 's' are transformed coords
+            a = self.rotation * np.pi / 180.
+            X = za_array * np.cos(az_array)
+            Y = za_array * np.sin(az_array)
+            Xs = (X * np.cos(a) - Y * np.sin(a)) / self.xstretch
+            Ys = (X * np.sin(a) + Y * np.cos(a)) / self.ystretch
+            
+            # Updated polar coordinates
+            theta_s = np.sqrt(Xs**2. + Ys**2.)
+            phi_s = np.arccos(Xs / theta_s)
+            phi_s[Ys < 0.] *= -1.
+            
+            # Fix coordinates below the horizon of the unstretched beam
+            theta_s[np.where(theta_s < 0.)] = 0.5 * np.pi
+            theta_s[np.where(theta_s >= np.pi/2.)] = 0.5 * np.pi
+            
+            # Update za_array and az_array
+            args = (phi_s, theta_s, freq_array)
+            az_array, za_array = phi_s, theta_s
+        
+        # Call interp() method on parent class
+        interp_fn = super(PerturbedPolyBeam, self).interp
+        interp_data, interp_basis_vector = interp_fn(*args, **kwargs)
+        
+        # Smooth step function
+        step = 0.5 * (1. + np.tanh((za_array - self.mainlobe_width)
+                                   / self.transition_width))
+        
+        # Add sidelobe perturbations
+        if self.nmodes != 0:
+            # Build Fourier series
+            p = 0
+            f_fac = 2.*np.pi / (np.pi/2.) #  Fourier series with period pi/2
+            for n in range(self.nmodes):
+                p += self.perturb_coeffs[n] * np.sin(f_fac * n * za_array)
+            p /= (np.max(p) - np.min(p)) / 2.
+            
+            # Modulate primary beam by perturbation function
+            interp_data *= (1. + step * p * self.perturb_scale)
+        
+        # Add mainlobe stretch factor
+        if self.mainlobe_scale != 1.:
+            # Subtract and re-add Gaussian normalized to 1 at za = 0
+            w = self.mainlobe_width / 2.
+            mainlobe0 = np.exp(-0.5*(za_array / w)**2.)
+            mainlobe_pert = np.exp(-0.5*(za_array/(w * self.mainlobe_scale))**2.)
+            interp_data += (1. - step) * (mainlobe_pert - mainlobe0)
+        
+        return interp_data, interp_basis_vector
+        
+    def beam_val(self, az_arr, za_arr, freqs, pol="XX"):
+        """
+        The entry point for healvis beam interpolation. Passes az_arr, za_arr, freqs,
+        to interp and returns the visibilities in the right shape.
+        """
+        if pol != "XX":
+            raise RuntimeError("Polybeam only does pol XX, not "+pol+". Called from healvis.")
+        interp_data, interp_basis_vector = self.interp(az_arr, za_arr, freqs)
+        return interp_data[0, 0, 1].T   # just want Npix, Nfreq
```

Comment:
Yes can be removed.

Reply:
Actually I'm going to remove beam_val() completely, because it can't be used with the current version of healvis, only a modified version we are using for sims.